From 6ea9d354f6144d72d993c1cd4fe139987f3c454e Mon Sep 17 00:00:00 2001
From: Greg Ungerer <gerg@kernel.org>
Date: Wed, 15 Mar 2023 23:45:48 +1000
Subject: [PATCH 3/3] elf2flt: force ARM.exidx section into text

Fix problem with el2flt core dumping on binutils version 2.33.1 and
newer. Previous attempts at fixing this have had a list of subtle
problems causing further breakage.

The problem stems from the .ARM.exidx section being marked as load
readonly data, but the elf2flt.ld linker script puts it into the
flat text section. The elf2flt processing code expects all separate
ELF sections marked as "data" to be in the flat data section and
calculates sizing and memory based on that. (Note that traditional
readonly data is treated as text in the linker script, so it is not
affected here and is merged into the flat text section with no
problems).

The .ARM.exidx section must be part of the flat text section - due
to the relocations that it carries. So move it into the text section
proper in the elf2flt.ld linker script.

Signed-off-by: Greg Ungerer <gerg@kernel.org>
---
 elf2flt.ld.in | 21 +++++++++------------
 1 file changed, 9 insertions(+), 12 deletions(-)

diff --git a/elf2flt.ld.in b/elf2flt.ld.in
index 0df999d..d86c0ba 100644
--- a/elf2flt.ld.in
+++ b/elf2flt.ld.in
@@ -76,6 +76,12 @@ W_RODAT:	*(.gnu.linkonce.r*)
 
 		/* .ARM.extab name sections containing exception unwinding information */
 		*(.ARM.extab* .gnu.linkonce.armextab.*)
+
+		/* .ARM.exidx name sections containing index entries for section unwinding */
+		@SYMBOL_PREFIX@__exidx_start = .;
+		*(.ARM.exidx* .gnu.linkonce.armexidx.*)
+		@SYMBOL_PREFIX@__exidx_end = .;
+
 		/* This is special code area at the end of the normal
 		   text section.  It contains a small lookup table at
 		   the start followed by the code pointed to by entries
@@ -84,19 +90,10 @@ W_RODAT:	*(.gnu.linkonce.r*)
 		PROVIDE(@SYMBOL_PREFIX@__ctbp = .);
 		*(.call_table_data)
 		*(.call_table_text)
-	} > flatmem :text
 
-	/* .ARM.exidx name sections containing index entries for section unwinding */
-	/* .ARM.exidx is sorted, so has to go in its own output section.  */
-	@SYMBOL_PREFIX@__exidx_start = .;
-	.ARM.exidx :
-	{
-		*(.ARM.exidx* .gnu.linkonce.armexidx.*)
-	} > flatmem
-	@SYMBOL_PREFIX@__exidx_end = .;
-
-	. = ALIGN(0x20) ;
-	@SYMBOL_PREFIX@_etext = . ;
+		. = ALIGN(0x20) ;
+		@SYMBOL_PREFIX@_etext = . ;
+	} > flatmem :text
 
 	.data : {
 		. = ALIGN(0x4) ;
-- 
2.25.1

